sources/ghostdog/src/infiniband.rs

```diff
@@ -0,0 +1,197 @@
+use crate::error;
+use crate::error::Result;
+use snafu::{ensure, ResultExt};
+use std::ffi::OsString;
+use std::ops::BitAnd;
+use std::path::Path;
+use std::str::FromStr;
+use std::{fs, str};
+
+const SYS_CLASS_DIR: &str = "/sys/class";
+const SW_MNG_CLASS: &str = "infiniband";
+const SW_MNG_VALUE: &str = "SW_MNG";
+
+/// Generic device struct found from sysfs
+#[derive(Clone, Debug)]
+pub(crate) struct SysDevice {
+    pub(crate) name: OsString,
+}
+
+/// Finds all the devices in sysfs for a given class sorted by name
+pub(crate) fn find_devices_by_class(class: String) -> Result<Vec<SysDevice>> {
+    let devices_path = Path::new(SYS_CLASS_DIR).join(class);
+    let mut found_devices: Vec<SysDevice> = vec![];
+    if !devices_path.exists() {
+        // Nothing to do since no devices detected
+        return Ok(found_devices);
+    }
+
+    let sys_devices_dirs = fs::read_dir(devices_path).context(error::InfinibandSysDevicesSnafu)?;
+    for device in sys_devices_dirs {
+        let device = device.context(error::InfinibandDeviceSnafu)?;
+        let device_name: OsString = device.file_name();
+        found_devices.push(SysDevice { name: device_name })
+    }
+
+    found_devices.sort_by_key(|device| device.name.clone());
+    Ok(found_devices)
+}
+
+/// Returns true if the SW_MNG_VALUE is found in the Vital Product Data file for given sysfs device
+pub(crate) fn is_device_sw_mng(device: OsString) -> Result<bool> {
+    let vpd_file = Path::new(SYS_CLASS_DIR)
+        .join(SW_MNG_CLASS)
+        .join(device)
+        .join("device")
+        .join("vpd");
+    if vpd_file.exists() {
+        let vpd_file_bytes: &[u8] =
+            &std::fs::read(vpd_file.as_path()).context(error::ReadFileSnafu {
+                path: vpd_file.as_path(),
+            })?;
+        let needle = SW_MNG_VALUE.as_bytes();
+        if vpd_file_bytes
+            .windows(needle.len())
+            .any(|window| window == needle)
+        {
+            return Ok(true);
+        }
+    }
+    Ok(false)
+}
+
+/// Struct that holds the two important values for an Infiniband Port when checking for
+/// Subnet Management capabilties.
+#[derive(Clone, Debug)]
+pub(crate) struct InfinibandPort {
+    /// Capability Mask is read in as a string such as 0xa751e848
+    pub(crate) capability_mask: CapabilityMask,
+    /// Port GUID is a u64 that is read in as a : formated string but written out in a form such as 0x
+    pub(crate) port_guid: Guid,
+}
+
+impl InfinibandPort {
+    /// is_sm_enabled checks the specific capability bit for Subnet Management means its enabled
+    // The actual bit means `isSMDisabled`, so check if its not zero
+    pub(crate) fn is_sm_enabled(&self) -> bool {
+        let bit_position: u32 = 10;
+        let mask: u32 = 1 << bit_position;
+        if !(&self.capability_mask & mask) != 0 {
+            return true;
+        }
+        false
+    }
```

Comment:
Negating the value after masking means that it will always be non-zero, because the zero bits in the mask get flipped to ones in the final value.

```suggestion
    pub(crate) fn is_sm_enabled(&self) -> bool {
        // SM is enabled if the eleventh bit ("SM disabled") is clear
        (&self.capability_mask & (1 << 10)) == 0
    }
```

Reply:
Thanks for catching this, I had ment to go back and confirm if this was going to work as expected.